### ISSUE TITLE
Add Support for Optional Bean Properties

### ISSUE THREAD

**Issue by karthicks** [typesafehub/config#399]
Currently, a config bean cannot define a property which is optional, in the sense that it need not necessarily have a value. This prevents us from using the `ConfigBeanFactory` for such config beans, where the optional property is actually supposed to be `null`.

To address this use case, I propose that we annotate the optional property as such, so that the bean factory does not raise a validation error, in the event that property happens to be missing. Specifically, let's define an annotation called `Optional` to indicate whether the annotated field is optional. In general, one can use annotations to suppress or express all kinds of validation rules, such as whether the property is an `Email` or a `PhoneNumber`, and so on.

To illustrate this, consider the following bean, which contains an optional property called `optionalValue`.

```
    public static class ValueObject {
        @Optional
        private String optionalValue;
        private String mandatoryValue;
    }
    private ValueObject valueObject;
```

One should be able to configure this object like so:

```
    "objects" : {
      "valueObject": {
        "mandatoryValue": "someValue"
      }
    }
```

Karthick Sankarachary

**Comment by havocp** [comment]
what about using the Optional class?

**Comment by karthicks** [comment]
Makes sense. Renamed the `Nullable` class to `Optional`, in the above pull request.

**Comment by havocp** [comment]
I meant to point to https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html (rather than using null). Then no annotation is needed.

**Comment by havocp** [comment]
you may find the two bugs linked from https://github.com/typesafehub/config/pull/286 have relevant discussion (I'm not sure, can't look in detail right now)

**Comment by karthicks** [comment]
You're right in that if we use `Optional`, then no annotation would be needed. The downside is that the bean won't be `Serializable` anymore. If fact, [this post ](http://stackoverflow.com/questions/24547673/why-java-util-optional-is-not-serializable-how-to-serialize-the-object-with-suc?answertab=active#tab-top) argues that:
> The primary design goal for `Optional` is to be used as the return value of functions when a return value might be absent. It was never intended for it to be used in other ways, such as for optional method arguments or to be stored as a field in an object.

I sort of agree with that, especially because in my use case, the config bean is part of a [GraphQL schema](http://graphql.org/docs/api-reference-type-system/#graphqlschema), which does not like `Optional` fields.

As far as #286 is concerned, I did consider using that approach, but unfortunately it doesn't work in the context of the `ContextBeanFactory`. In other words, we can't leverage that feature when trying to convert a config object to a config bean, as far as I can tell.

In Scala, the `Option` class is apparently `Serializable`. Again, I'm not sure if that will play well with GraphQL (and similar data models for client-server applications).

Last, if we take the annotation approach, then we could potentially use that for defining custom types such as emails and phone numbers, etc.

**Comment by karthicks** [comment]
Just FYI, I got rid of the dependencies I had added in `config/build.sbt`, to address @viktorklang's comment in the pull request. Let me know if the annotation approach will work, or whether an alternative is preferred.

**Comment by karthicks** [comment]
@havocp, can we close this issue, given that the pull request was merged, and is now part of v1.3.1?

**Comment by havocp** [comment]
Sure!
